jgrapht-core/src/main/java/org/jgrapht/graph/AsGraphUnion.java

```diff
@@ -0,0 +1,459 @@
+/*
+ * (C) Copyright 2009-2017, by Ilya Razenshteyn and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.graph;
+
+import java.io.*;
+import java.util.*;
+
+import org.jgrapht.*;
+import org.jgrapht.util.*;
+
+/**
+ * Read-only union of two graphs.
+ * 
+ * <p>
+ * Read-only union of two graphs: G<sub>1</sub> and G<sub>2</sub>. If G<sub>1</sub> =
+ * (V<sub>1</sub>, E<sub>1</sub>) and G<sub>2</sub> = (V<sub>2</sub>, E<sub>2</sub>) then their
+ * union G = (V, E), where V is the union of V<sub>1</sub> and V<sub>2</sub>, and E is the union of
+ * E<sub>1</sub> and E<sub>1</sub>. A {@link WeightCombiner} in order to calculate edge weights.
+ * 
+ * @param <V> the vertex type
+ * @param <E> the edge type
+ * 
+ * @author Ilya Razenshteyn
+ */
+public class AsGraphUnion<V, E>
+    extends AbstractGraph<V, E>
+    implements Serializable
+{
+    private static final long serialVersionUID = -740199233080172450L;
+
+    private static final String READ_ONLY = "union of graphs is read-only";
+
+    private final Graph<V, E> g1;
+    private final GraphType type1;
+    private final Graph<V, E> g2;
+    private final GraphType type2;
+    private final GraphType type;
+    private final WeightCombiner operator;
+
+    /**
+     * Construct a new graph union.
+     * 
+     * @param g1 the first graph
+     * @param g2 the second graph
+     * @param operator the weight combiner (policy for edge weight calculation)
+     */
+    public AsGraphUnion(Graph<V, E> g1, Graph<V, E> g2, WeightCombiner operator)
+    {
+        this.g1 = GraphTests.requireDirectedOrUndirected(g1);
+        this.type1 = g1.getType();
+
+        this.g2 = GraphTests.requireDirectedOrUndirected(g2);
+        this.type2 = g2.getType();
+
+        if (g1 == g2) {
+            throw new IllegalArgumentException("g1 is equal to g2");
+        }
+        this.operator = Objects.requireNonNull(operator, "Weight combiner cannot be null");
+
+        // compute result type
+        GraphType t = DefaultGraphType.mixed();
```

Comment:
We should probably add some more information to the GraphType of the uniongraph, e.g. allowSelfLoops=true if either of the base graphs allows them, isAllowingMultipleEdges=true by default (if the base graphs g1 and g2 both have an edge between vertices u,v then getAllEdges() returns both edges),